trin-cli/src/main.rs

```diff
@@ -0,0 +1,137 @@
+use std::os::unix::net::UnixStream;
+use std::path::PathBuf;
+use structopt::StructOpt;
+use thiserror::Error;
+
+use trin_core::cli::DEFAULT_WEB3_IPC_PATH;
+
+#[derive(StructOpt, Debug)]
+#[structopt(
+    name = "trin-cli",
+    version = "0.0.1",
+    author = "Ethereum Foundation",
+    about = "Run JSON-RPC commands against trin nodes"
+)]
+struct Config {
+    #[structopt(
+        default_value(DEFAULT_WEB3_IPC_PATH),
+        long,
+        help = "path to JSON-RPC endpoint"
+    )]
+    ipc: PathBuf,
+
+    #[structopt(help = "e.g. discv5_routingTableInfo", required = true)]
+    endpoint: String,
+}
+
+fn main() {
+    let Config { ipc, endpoint } = Config::from_args();
+
+    eprintln!(
+        "Attempting RPC. endpoint={} file={}",
+        endpoint,
+        ipc.to_string_lossy()
+    );
+    let mut client = match TrinClient::from_ipc(&ipc) {
+        Ok(client) => client,
+        Err(err) => {
+            eprintln!("Could not connect. err={:?}", err);
+            return;
+        }
+    };
+
+    let req = client.build_request(endpoint.as_str());
+    let resp = client.make_request(req);
+
+    match resp {
+        Err(error) => {
+            eprintln!("error: {}", error);
+        }
+        Ok(value) => {
+            // unwrap: this value is safe to serialize, it was just deserialized!
+            println!("{}", serde_json::to_string_pretty(&value).unwrap());
+        }
+    }
+}
+
+fn build_request<'a>(method: &'a str, request_id: u64) -> jsonrpc::Request<'a> {
+    jsonrpc::Request {
+        method: method,
+        params: &[],
+        id: serde_json::json!(request_id),
+        jsonrpc: Some("2.0"),
+    }
+}
+
+pub trait TryClone {
+    fn try_clone(&self) -> std::io::Result<Self>
+    where
+        Self: Sized;
+}
+
+impl TryClone for UnixStream {
+    fn try_clone(&self) -> std::io::Result<Self> {
+        UnixStream::try_clone(self)
+    }
+}
+
+pub struct TrinClient<S>
+where
+    S: std::io::Read + std::io::Write + TryClone,
+{
+    stream: S,
+    request_id: u64,
+}
+
+impl TrinClient<UnixStream> {
+    fn from_ipc(path: &PathBuf) -> std::io::Result<Self> {
+        // TODO: a nice error if this file does not exist
+        Ok(Self {
+            stream: UnixStream::connect(path)?,
+            request_id: 0,
+        })
+    }
+}
+
+#[derive(Error, Debug)]
+pub enum JsonRpcError {
+    #[error("Received malformed response: {0}")]
+    Malformed(serde_json::Error),
+
+    #[error("Received empty response")]
+    Empty,
+}
+
+// TryClone is used because JSON-RPC responses are not followed by EOF. We must read bytes
+// from the stream until a complete object is detected, and the simplest way of doing that
+// with available APIs is to give ownership of a Read to a serde_json::Deserializer. If we
+// gave it exclusive ownership that would require us to open a new connection for every
+// command we wanted to send! By making a clone (or, by trying to) we can have our cake
```

Comment:
I guess that's only important with an interactive REPL-type tool, rather than this CLI one-shot, right?

Reply:
Yeah, definitely. I can drop it if you would prefer. This was originally added because I had grand plans of turning this into a client we could use in other parts of the codebase which want to make JSON-RPC calls, such as in the peer tester. You're right that it's unnecessary for current usage but I still think it's useful to have that long-term goal for `TrinClient`.

Reply:
Yeah, I'm happy to keep it. Maybe just add that little bit of context to the comment.